pkg/controllers/jobset_controller.go

```diff
@@ -30,30 +37,272 @@ type JobSetReconciler struct {
+		} else {
+			klog.V(0).Info("deleted old failed job", "job", job)
+		}
+	}
+}
+
+// getChildJobs fetches all Jobs owned by the JobSet and returns them
+// categorized by status (active, successful, failed).
+func (r *JobSetReconciler) getChildJobs(ctx context.Context, js *jobset.JobSet, req ctrl.Request) (*childJobs, error) {
+	// Get all active jobs owned by JobSet.
+	var childJobList batchv1.JobList
+	if err := r.List(ctx, &childJobList, client.InNamespace(req.Namespace), client.MatchingFields{jobOwnerKey: req.Name}); err != nil {
+		klog.Error(err, "unable to list child Jobs")
+		return nil, err
+	}
+
+	jobs := childJobs{}
+	for i, job := range childJobList.Items {
+		_, finishedType := IsJobFinished(&job)
+		switch finishedType {
+		case "": // active
+			jobs.active = append(jobs.active, &childJobList.Items[i])
+		case batchv1.JobFailed:
+			jobs.failed = append(jobs.failed, &childJobList.Items[i])
+		case batchv1.JobComplete:
+			jobs.successful = append(jobs.successful, &childJobList.Items[i])
+		}
+	}
+
+	return &jobs, nil
+}
+
+func (r *JobSetReconciler) createReadyJobs(ctx context.Context, js *jobset.JobSet) error {
+	for _, jobTemplate := range js.Spec.Jobs {
+		job, err := r.constructJobFromTemplate(js, &jobTemplate)
+		if err != nil {
+			klog.Error(err, "unable to construct job from template", "jobTemplate", jobTemplate)
+			return err
+		}
+
+		// Check if we need to create this job.
+		// If not, skip this job and continue iterating to the next job.
+		create, err := r.shouldCreateJob(ctx, job)
+		if err != nil {
+			return err
+		}
+		if !create {
+			klog.Infof("skipping job %s", job.Name)
+			continue
+		}
+
+		// Create headless service if specified for this job.
+		if jobTemplate.Network.EnableDNSHostnames != nil && *jobTemplate.Network.EnableDNSHostnames {
+			klog.Infof("creating headless service: %s", job.Name)
+			if err := r.createHeadlessSvcIfNotExist(ctx, js, job); err != nil {
+				klog.Infof("error creating headless service: %v", err)
+				return err
+			}
+		}
+
+		// Create the job.
+		klog.Infof("creating job %s", job.Name)
+		if err := r.Create(ctx, job); err != nil {
+			klog.Error(err, "unable to create Job for JobSet", "job", job)
+			return err
+		}
+	}
+	return nil
+}
+
+// TODO: look into adopting service and updating the selector
+// if it is not matching the job selector.
+func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, js *jobset.JobSet, job *batchv1.Job) error {
+	// Check if service already exists. Service name is same as job name.
+	// If the service does not exist, create it.
+	var headlessSvc corev1.Service
+	if err := r.Get(ctx, types.NamespacedName{Name: job.Name, Namespace: js.Namespace}, &headlessSvc); err != nil {
+		headlessSvc := corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      job.Name,
+				Namespace: js.Namespace,
+			},
+			Spec: corev1.ServiceSpec{
+				ClusterIP: "None",
+				Selector:  genLabelSelector(job),
```

Comment:
Sorry for the back and forth, we actually can use the job-name label because services are namespaced and they will only look at pods within the same namespace where the job is.